docs/proposals/hosted_components_templates.md

```diff
@@ -0,0 +1,559 @@
+
+## Description
+Currently most of our hosted components have their own repositories where they
+maintain changes and then need to synchronize those changes within their
+respective roles. This can cause delays and times where features don't
+match up with what is installed or can be configured.
+
+There is also a concern that copying files from one location to another can
+cause cases where files are out of sync or versions don't align with what is
+installed and what is configured (is what we configured correct for the image
+version we installed).
+
+## Rationale
+As was piloted with the Template Service Broker ("TSB" for short), a role was
+created to process and configure OCP templates that were synchronized (currently
+still a manual process) from the origin repo where the component resides.
+
+This allowed any changes for the installation to be maintained by the developers
+and keep them from needing to be fluent with Ansible. It also afforded a level
+of abstraction from needing to keep in sync with TSB related changes and then
+translate them to changes within an Ansible role.
+
+However, there is still the gap of ensuring a role configures changes appropriately
+for the version of the component it is installing and we are violating the "DRY"
+principle by needing to sync files between repos.
+
+Prior to 3.4, Aggregated Logging solved this using a deployer pod. Everything
+was versioned and contained within this artifact that the Integration
+Services team provided and maintained. The deployer pod lived in the same
+repo as the components that it configured. This is something that is addressed
+again now with the Service Broker, Ansible Playbook Bundles ("APBs" for short)
+are configured in such a way that a containerized install/uninstall is done for
+the component.
+
+## Design
+We should describe a way to create a containerized playbook and role to mirror
+what is done with APBs for hosted components so that the openshift-ansible roles
+would reuse these containers to install/uninstall these components.
+
+The containerized role could follow the pattern used by the TSB within its
+openshift-ansible role `template_service_broker` to maintain a lower learning
+curve for those not familiar with ansible as well as decrease duplication of work
+in cases where the component provides OCP templates already.
+
+Within the openshift-ansible role, installing in a way similar to APBs would
+create an interface that would not need to change once the role is created,
+with the exception being if new information need be passed into the container.
+
+If hosted components adhere to the structure required for the Service Broker
+to provision/deprovision bind/unbind it allows us to also open up the possibility
+to set up these components through the UI. This could further lighten the load
+placed on openshift-ansible and provide an alternative means to install hosted
+components after an OCP installation is completed.
```

Comment:
We just had a call today with the APB devs and they mentioned some future goals to have a cluster admin like interface where admins only could see options to install optional add-on applications like this after the cluster was up.